context.py

```diff
@@ -0,0 +1,61 @@
+import abc
+from datetime import datetime
+from typing import Any, Dict, Optional
+
+class ContextStorageEngine(abc.ABC):
+    @abc.abstractmethod
+    def get(self, key: str, context_type: str) -> Optional[Dict[str, Any]]:
+        pass
+
+    @abc.abstractmethod
+    def set(self, key: str, value: Any, context_type: str) -> None:
+        pass
+
+    @abc.abstractmethod
+    def delete(self, key: str, context_type: str) -> None:
+        pass
+
+class InMemoryContextStorage(ContextStorageEngine):
+    def __init__(self):
+        self.contact_contexts = {}
+        self.thread_contexts = {}
+
+    def get(self, key: str, context_type: str) -> Optional[Dict[str, Any]]:
+        contexts = self.contact_contexts if context_type == 'contact' else self.thread_contexts
+        return contexts.get(key)
+
+    def set(self, key: str, value: Any, context_type: str) -> None:
+        contexts = self.contact_contexts if context_type == 'contact' else self.thread_contexts
+        if key not in contexts:
+            contexts[key] = {}
+        contexts[key]['value'] = value
+        contexts[key]['timestamp'] = datetime.now().isoformat()
+
+    def delete(self, key: str, context_type: str) -> None:
+        contexts = self.contact_contexts if context_type == 'contact' else self.thread_contexts
+        if key in contexts:
+            del contexts[key]
+
+class Context:
```

Comment:
Unify the getting and setting functions so that the default is thread, and an optional argument will cause it to be set or loaded from the contact. The get function should default to thread, but will also return the value from the contact if it exists. Getting a value returns an object that contains the value, the source (context or thread), and the timestamp it was set.